proposals/003-authentication-api.md

```diff
@@ -0,0 +1,650 @@
+
+```java
+    /**
+     * @return The TLS context for the server connection, or empty if the server connection is not TLS.
+     */
+    Optional<ServerTlsContext> serverTlsContext();
+```
+
+Where
+
+```java
+package io.kroxylicious.proxy.tls;
+
+import java.security.cert.X509Certificate;
+import java.util.Optional;
+
+public interface ServerTlsContext {
+    /**
+     * @return The TLS server certificate that the proxy presented to the server during TLS handshake,
+     * or empty if no TLS client certificate was presented during TLS handshake.
+     */
+    Optional<X509Certificate> proxyClientCertificate();
+
+    /**
+     * @return the server's TLS certificate.
+     */
+    X509Certificate serverCertificate();
+}
+```
+
+### APIs for Filters to produce server SASL information
+
+A SASL Initiator will be able to use the follow methods on `FilterContext` to announce a successful, or failed, authentication with a Kafka server:
+
+```java
+    /**
+     * Allows a filter
+     * to announce a successful authentication outcome with the Kafka server to other plugins.
+     * After calling this method the result of {@link #serverSaslContext()} will
+     * be non-empty for this and other filters.
+     * This method may be called multiple times over the lifetime of
+     * a session if reauthentication is required.
+     * TODO define the semantics around reauth
+     * @param saslPrincipal The authenticated principal.
+     */
+    void serverSaslAuthenticationSuccess(String mechanism, String serverName);
+
+    /**
+     * Allows a filter
+     * to announce a failed authentication outcome with the Kafka server.
+     * @param exception An exception describing the authentication failure.
+     */
+    void serverSaslAuthenticationFailure(String mechanism, String serverName, Exception exception);
+```
+
+### APIs for Filters to consume server SASL information
+
+```java
+    /**
+     * @return The SASL context for the server connection, or empty if the server
+     * has not successfully authenticated using SASL.
+     */
+    Optional<ServerSaslContext> serverSaslContext();
+```
+
+Where:
+
+```java
+package io.kroxylicious.proxy.authentication;
+
+import java.util.Optional;
+
+import io.kroxylicious.proxy.filter.FilterContext;
+
+/**
+ * Exposes SASL authentication information to plugins, for example using {@link FilterContext#serverSaslContext()} ()}.
+ * This is implemented by the runtime for use by plugins.
+ */
+public interface ServerSaslContext {
+
+    /**
+     * The name of the SASL mechanism used.
+     * @return The name of the SASL mechanism used.
+     */
+    String mechanismName();
+
+    /**
+     * Returns the principal returned by the server.
+     * @return the principal returned by the server,
+     * or empty if the SASL mechanism used does not support mutual authentication.
```

Comment:
I didn't realise SASL mutual auth was a thing.  What mechanisms support it?

Reply:
Why do we still have distinct interfaces for ClientSaslContext and ServerSaslContext.  I think we can have a single SaslContext used for both the client (inspector) and server (initiator) use cases.  There is no reason I can think of to hide the `authorizationId` that the initiator has used.